nexus_transporter/src/TransporterNode.cpp

```diff
@@ -459,44 +453,72 @@ TransporterNode::TransporterNode(const rclcpp::NodeOptions& options)
   );
 }
 
-bool TransporterNode::registration_callback()
+void TransporterNode::Data::_register()
 {
-  RCLCPP_INFO(this->get_logger(), "Registering with system orchestrator...");
-  auto client = this->create_client<RegisterTransporter>(
-    RegisterTransporterService::service_name(),
-    rclcpp::ServicesQoS(),
-    _data->cb_group
-  );
-
-  if (!client->wait_for_service(_data->connection_timeout))
+  auto node = w_node.lock();
+  if (node == nullptr)
   {
-    RCLCPP_ERROR(this->get_logger(), "Could not find system orchestrator!");
-    return false;
+    return;
   }
 
-  auto req = std::make_shared<RegisterTransporter::Request>();
-  req->description.workcell_id = this->get_name();
-  auto fut = client->async_send_request(req);
-
-  if (fut.wait_for(_data->connection_timeout) != std::future_status::ready)
+  if (this->ongoing_register)
   {
     RCLCPP_ERROR(
-      this->get_logger(), "Could not register with system orchestrator");
-    return false;
+      node->get_logger(),
+      "Failed to register: No response from system orchestrator.");
+    if (!this->register_client->remove_pending_request(*this->
+      ongoing_register))
+    {
+      RCLCPP_WARN(node->get_logger(),
+        "Unable to remove pending request during transporter registration.");
+    }
   }
-  auto resp = fut.get();
-  if (!resp->success)
+
+  RCLCPP_INFO(node->get_logger(), "Registering with system orchestrator...");
+  auto register_cb =
+    [this, node](rclcpp::Client<RegisterTransporter>::SharedFuture future)
```

Comment:
we should capture `w_node` instead, and lock it inside `register_cb` since this is an async request and `node` is expected to be released as `_register` returns before we get a response